### DISCUSSION TITLE
How can JsonSerializer.Deserialize ever return null?

### DISCUSSION THREAD

**Discussion by KieranDevvs** [dotnet/runtime#77317] [Q&A] [answered]
All of the overload methods for `Deserialize<T>` return `T?` but in my testing:

- If the JSON is invalid, it throws.
- If the JSON is valid but doesn't match the type, `T`'s constructor is called but no members are populated (thus `T` is not null).
- If the JSON is `string.Empty` or `null`, it throws.
- If the JSON is valid and all fields can be deserialized then `T` is not null and fully populated.
- If the JSON is valid and *some* fields can be deseralized then `T` is not null and is partially populated.

I ask because in my code I'm doing the relevant type checks i.e
```cs
if (deserializedObject is null)
{
    // deal with T?...
}
```

and when unit testing this code, I cant get a scenario where I can meet this condition.

It then begs the question, if `T` cannot be null (or maybe it can and someone can elaborate for me), then why does the signature return `T?`

I have a feeling it might be to do with some JSON serialization options that aren't set by default that would return null rather than throwing in some cases, but I cant find anything that resembles this behavior in my own findings.

**Comment by gfoidl** [comment, marked as answer]
E.g.
```c#
using System.Text.Json;

object? obj = JsonSerializer.Deserialize<object>("null");
Console.WriteLine(obj is null);
```
